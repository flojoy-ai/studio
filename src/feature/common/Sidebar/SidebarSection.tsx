import { useState } from "react";
import {
  createStyles,
  UnstyledButton,
  Group,
  Box,
  Collapse,
} from "@mantine/core";
import { IconChevronLeft, IconChevronRight } from "@tabler/icons-react";
import { Children } from "react";

export const useSidebarStyles = createStyles((theme) => ({
  control: {
    fontWeight: 500,
    display: "block",
    width: "90%",
    padding: `${theme.spacing.xs} ${theme.spacing.xs}`,
    color: "black",
    fontSize: theme.fontSizes.sm,
    margin: "0px 20px 5px 20px",

    borderRadius: 2,
    backgroundColor: theme.colors.accent1[0],
  },
  title: {
    color:
      theme.colorScheme === "dark"
        ? theme.colors.dark[6]
        : theme.colors.gray[0],
  },

  chevron: {
    transition: "transform 200ms ease",
    color:
      theme.colorScheme === "dark"
        ? theme.colors.dark[6]
        : theme.colors.gray[0],
  },
}));

type SidebarSectionProps = {
  title: string;
  depth: number;
  children: React.ReactNode;
};

const SidebarSection = ({ depth, title, children }: SidebarSectionProps) => {
  const [opened, setOpened] = useState(false);
  const { classes, theme } = useSidebarStyles();
  const ChevronIcon = theme.dir === "ltr" ? IconChevronRight : IconChevronLeft;

  if (Children.toArray(children).every((child) => child === null)) {
    return null;
  }

  return (
    <>
      <UnstyledButton
        onClick={() => setOpened((o) => !o)}
        className={classes.control}
        data-cy="sidebar-section-btn"
      >
        <Group position="apart" spacing={0}>
          <Box sx={{ display: "flex", alignItems: "center" }}>
            <Box ml="md" className={classes.title}>
              {title}
            </Box>
          </Box>
          <ChevronIcon
            className={classes.chevron}
            size={14}
            stroke="1.5"
            style={{
              transform: opened
                ? `rotate(${theme.dir === "rtl" ? -90 : 90}deg)`
                : "none",
            }}
          />
        </Group>
      </UnstyledButton>
      <Collapse in={opened}>
        {/* padding according to the depth of the section */}
        <div style={{ paddingLeft: `${10 + (depth + 1) * 8}px` }}>
          {children}
        </div>
      </Collapse>
    </>
  );
};
export default SidebarSection;
