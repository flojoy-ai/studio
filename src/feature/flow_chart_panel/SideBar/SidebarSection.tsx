import { useState, useCallback } from "react";
import {
  Group,
  Box,
  Collapse,
  Text,
  UnstyledButton,
  createStyles,
} from "@mantine/core";
import { IconChevronLeft, IconChevronRight } from "@tabler/icons";
import SidebarSubSection from "./SidebarSubSection";
import { useFlowChartState } from "@src/hooks/useFlowChartState";
import { NodeOnAddFunc, ParamTypes } from "../types/NodeAddFunc";
import { v4 as uuidv4 } from "uuid";
import { ElementsData } from "../types/CustomNodeProps";

const useStyles = createStyles((theme) => ({
  control: {
    fontWeight: 500,
    display: "block",
    width: "90%",
    padding: `${theme.spacing.xs}px ${theme.spacing.md}px`,
    color: "black",
    fontSize: theme.fontSizes.sm,
    margin: "10px 20px 10px 20px",
    backgroundColor: theme.colorScheme === "dark" ? "#94F4FC" : "#E1E4E7",
  },

  subSection: {
    fontWeight: 500,
    display: "block",
    textDecoration: "none",
    padding: `${theme.spacing.xs}px ${theme.spacing.md}px`,
    paddingLeft: 31,
    marginLeft: 30,
    fontSize: theme.fontSizes.sm,
    color:
      theme.colorScheme === "dark"
        ? theme.colors.dark[0]
        : theme.colors.gray[7],
    borderLeft: `1px solid ${
      theme.colorScheme === "dark" ? theme.colors.dark[4] : theme.colors.gray[3]
    }`,
  },

  chevron: {
    transition: "transform 200ms ease",
  },
}));

interface LinksGroupProps {
  title: string;
  child?: { name: string; key: string }[];
}

const getNodePosition = () => {
  return {
    x: 50 + Math.random() * 20,
    y: 50 + Math.random() + Math.random() * 20,
  };
};

const SidebarSection = ({ title, child }: LinksGroupProps) => {
  const { classes, theme } = useStyles();
  const hasChilds = Array.isArray(child);
  const [opened, setOpened] = useState(false);
  const ChevronIcon = theme.dir === "ltr" ? IconChevronRight : IconChevronLeft;

  const { nodes, setNodes } = useFlowChartState();

  const addNewNode: NodeOnAddFunc = useCallback(
    ({ funcName, params, type, inputs, uiComponentId }) => {
      let nodeLabel: string;
      const nodeId = `${funcName}-${uuidv4()}`;
      if (funcName === "CONSTANT") {
        nodeLabel = "2.0";
      } else {
        const numOfThisNodesOnChart = nodes.filter(
          (node) => node.data.func === funcName
        ).length;
        nodeLabel =
          numOfThisNodesOnChart > 0
            ? `${funcName}_${numOfThisNodesOnChart}`
            : funcName;
      }
      const nodeParams = params
        ? Object.keys(params).reduce(
            (prev: ElementsData["ctrls"], param) => ({
              ...prev,
              [param]: {
                functionName: funcName,
                param,
                value:
                  funcName === "CONSTANT"
                    ? nodeLabel
                    : params![param].default?.toString(),
              },
            }),
            {}
          )
        : {};

      const newNode = {
        id: nodeId,
        type: uiComponentId || type,
        data: {
          id: nodeId,
          label: nodeLabel,
          func: funcName,
          type,
          ctrls: nodeParams,
          inputs,
        },
        position: getNodePosition(),
      };
      setNodes((els) => els.concat(newNode));
    },
    [nodes, setNodes]
  );

  const items = (hasChilds ? child : []).map((c) => (
    <Text<"div"> component="div" className={classes.subSection} key={c.name}>
      <SidebarSubSection subSection={c} onAdd={addNewNode} />
    </Text>
  ));

  return (
    <>
      <UnstyledButton
        onClick={() => setOpened((o) => !o)}
        className={classes.control}
      >
        <Group position="apart" spacing={0}>
          <Box sx={{ display: "flex", alignItems: "center" }}>
            <Box ml="md">{title}</Box>
          </Box>
          {hasChilds && (
            <ChevronIcon
              className={classes.chevron}
              size={14}
              stroke={1.5}
              style={{
                transform: opened
                  ? `rotate(${theme.dir === "rtl" ? -90 : 90}deg)`
                  : "none",
              }}
            />
          )}
        </Group>
      </UnstyledButton>
      {hasChilds ? <Collapse in={opened}>{items}</Collapse> : null}
    </>
  );
};

export default SidebarSection;
