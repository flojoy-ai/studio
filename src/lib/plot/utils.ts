export function clamp(x: number, lo: number, hi: number) {
  return Math.min(Math.max(x, lo), hi);
}
