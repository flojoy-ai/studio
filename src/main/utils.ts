import * as http from "http";
import * as fs from "fs";
import { isIP } from "net";
import { execCommand } from "./executor";
import { Command } from "./command";
import { app, dialog } from "electron";
import { join } from "path";
import { killCaptain } from "./python";
import log from "electron-log";
import { ChildProcess } from "node:child_process";
import { Err, Result, Ok } from "../types/result";

export const isPortFree = (port: number) =>
  new Promise((resolve) => {
    const server = http
      .createServer()
      .listen(port, "127.0.0.1", () => {
        server.close();
        resolve(true);
      })
      .on("error", () => {
        resolve(false);
      });
  });

export const killProcess = async (port: number) => {
  await execCommand(
    new Command({
      darwin: `kill -9 $(lsof -t -i :${port})`,
      linux: `kill -9 $(lsof -t -i :${port})`,
      win32: `FOR /F "tokens=5" %i IN ('netstat -aon ^| find "${port}"') DO Taskkill /F /PID %i`,
    }),
  );
};

export const ping = async (addr: string) => {
  if (isIP(addr) === 0) {
    throw new Error(`Invalid IP address: ${addr}`);
  }

  return await execCommand(
    new Command({
      win32: `ping -n 1 ${addr}`,
      darwin: `ping -c 1 ${addr}`,
      linux: `ping -c 1 ${addr}`,
    }),
    { quiet: true },
  );
};

export const netstat = async () => {
  return await execCommand(
    new Command({
      win32: "netstat",
      darwin: "netstat",
      linux: "netstat",
    }),
    { quiet: true },
  );
};

export const ifconfig = async () => {
  return await execCommand(
    new Command({
      win32: "ipconfig /all",
      darwin: "ifconfig",
      linux: "ifconfig",
    }),
    { quiet: true },
  );
};

export const writeFileSync = (_, filePath: string, text: string): void => {
  fs.writeFileSync(filePath, text);
};

export const pickDirectory = async (): Promise<string> => {
  const handler = await dialog.showOpenDialog({
    properties: ["openDirectory"],
  });
  return handler.canceled ? "" : handler.filePaths[0];
};

export const getCustomBlocksDir = async () => {
  const filePath = join(app.getPath("home"), ".flojoy/custom_blocks_path.txt");

  if (!fs.existsSync(filePath)) {
    return undefined;
  }

  const blocksPath = fs
    .readFileSync(filePath, { encoding: "utf-8" })
    .toString();

  if (!fs.existsSync(blocksPath)) {
    return undefined;
  }

  return blocksPath;
};

export const cacheCustomBlocksDir = (_, dirPath: string) => {
  const flojoyDir = join(app.getPath("home"), ".flojoy");
  if (!fs.existsSync(flojoyDir)) {
    fs.mkdirSync(flojoyDir);
  }
  const cacheFilePath = join(flojoyDir, "custom_blocks_path.txt");
  fs.writeFileSync(cacheFilePath, dirPath);
};

export const openTestPicker = (): Promise<
  { filePath: string; fileContent: string } | undefined
> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: ["json", "py"],
            name: "Test",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const openFilePicker = (
  _,
  allowedExtensions: string[] = ["json"],
): Promise<{ filePath: string; fileContent: string } | undefined> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: allowedExtensions,
            name: "allowed extensions",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
        resolve(undefined);
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const cleanup = async () => {
  const captainProcess = global.captainProcess as ChildProcess;
  log.info(
    "Cleanup function invoked, is captain running? ",
    !(captainProcess?.killed ?? true),
  );
  if (captainProcess && captainProcess.exitCode === null) {
    const success = killCaptain();
    if (success) {
      global.captainProcess = null;
      log.info("Successfully terminated captain :)");
    } else {
      log.error("Something went wrong when terminating captain!");
    }
  }
};

export const loadFileFromFullPath = async (
  filePath: string,
): Promise<string> => {
  return fs.readFileSync(filePath, { encoding: "utf-8" }).toString();
};

export const saveFileToFullPath = async (
  filePath: string,
  content: string,
): Promise<Result<void>> => {
  try {
    fs.writeFileSync(filePath, content);
    return Ok(undefined);
  } catch (error) {
    return Err(error as Error);
  }
};
