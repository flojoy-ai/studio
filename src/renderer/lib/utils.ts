import { type ClassValue, clsx } from "clsx";
import { twMerge } from "tailwind-merge";
import { z } from "zod";
import * as E from "fp-ts/Either";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function tryParse<Z extends z.ZodTypeAny>(z: Z) {
  return (val: unknown) => {
    const res = z.safeParse(val);
    if (res.success) {
      return E.right(res.data as z.infer<Z>);
    }
    return E.left(res.error);
  };
}
