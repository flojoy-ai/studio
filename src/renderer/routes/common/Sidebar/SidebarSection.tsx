import { useEffect, useState } from "react";
import { Children } from "react";
import { twMerge } from "tailwind-merge";
import { cva } from "class-variance-authority";
import { categoryMap, sidebarVariants } from "./SidebarNode";
import {
  Collapsible,
  CollapsibleContent,
  CollapsibleTrigger,
} from "@src/components/ui/collapsible";

const bgOverrideVariants = cva(undefined, {
  variants: {
    variant: {
      DATA: "bg-accent2/10",
      ETL: "bg-accent1/10",
      IO: "bg-accent4/10",
      LOGIC: "bg-accent3/10",
      AUTOGEN: "bg-blue-500/10",
    },
  },
});

type SidebarSectionProps = {
  title: string;
  depth: number;
  children: React.ReactNode;
  expand: boolean;
  collapse: boolean;
  icon?: React.ReactNode;
  category?: string;
  autogeneratedCategory?: boolean;
};

const SidebarSection = ({
  depth,
  title,
  children,
  expand,
  collapse,
  category,
  icon,
  autogeneratedCategory,
}: SidebarSectionProps) => {
  const [opened, setOpened] = useState(false);

  useEffect(() => {
    setOpened(true);
  }, [expand]);

  useEffect(() => {
    setOpened(false);
  }, [collapse]);

  if (Children.toArray(children).every((child) => child === null)) {
    return null;
  }

  const variant = categoryMap[category ?? "TRANSFORMERS"] ?? "ETL";

  return (
    <Collapsible open={opened} onOpenChange={setOpened}>
      <CollapsibleTrigger
        className={twMerge(
          "mb-2 flex max-h-10 w-full items-center rounded-sm border px-2 py-2.5",
          sidebarVariants({ variant }),
          bgOverrideVariants({ variant }),
        )}
        data-cy="sidebar-section-btn"
      >
        {icon}
        <div
          className={twMerge(
            "ml-2 mr-auto font-medium tracking-wide",
            autogeneratedCategory ? "font-mono" : "",
          )}
        >
          {title}
        </div>
      </CollapsibleTrigger>
      <CollapsibleContent className="CollapsibleContent">
        {/* padding according to the depth of the section */}
        <div style={{ paddingLeft: `${(depth + 1) * 5}px` }}>{children}</div>
      </CollapsibleContent>
    </Collapsible>
  );
};
export default SidebarSection;
