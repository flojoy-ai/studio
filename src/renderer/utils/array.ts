export function filterMap<T, U>(
  arr: Array<T>,
  mapPred: (x: T) => U | undefined | null,
) {
  return arr.reduce((acc: Array<U>, x: T) => {
    const val = mapPred(x);
    if (val !== undefined && val !== null) {
      acc.push(val);
    }

    return acc;
  }, []);
}
