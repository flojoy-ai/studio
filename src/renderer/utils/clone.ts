import { DeepMutable } from "@/renderer/types/util";

export const deepMutableClone = <T>(obj: T): DeepMutable<T> => {
  return JSON.parse(JSON.stringify(obj)) as DeepMutable<T>;
};
