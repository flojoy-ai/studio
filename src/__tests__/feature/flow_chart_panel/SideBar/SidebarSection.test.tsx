import { fireEvent } from "@testing-library/react";
import SidebarSection from "@src/feature/flow_chart_panel/SideBar/SidebarSection";
import { renderWithTheme } from "@src/__tests__/__utils__/utils";

class ResizeObserver {
  observe() {}
  unobserve() {}
}

const items = {
  title: "test",
  child: [{ name: "test", key: "test" }],
};

jest.mock("@src/feature/flow_chart_panel/manifest/COMMANDS_MANIFEST", () => {
  return {
    COMMANDS: [],
    SECTIONS: [],
  };
});

jest.mock("@src/feature/flow_chart_panel/manifest/PARAMETERS_MANIFEST", () => {
  return {
    FUNCTION_PARAMETERS: {},
  };
});

jest.mock("@src/hooks/useFlowChartGraph", () => {
  return {
    useFlowChartGraph: () => {
      return {
        setNodes: jest.fn(),
      };
    },
  };
});

jest.mock("uuid", () => {
  return {
    v4: jest.fn(),
  };
});

describe("SidebarSection", () => {
  window.ResizeObserver = ResizeObserver as any;
  it("matches snapshot", () => {
    const { container } = renderWithTheme(
      <SidebarSection
        title={items.title}
        child={items.child}
        setNodes={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });
  it("checks the section button in document", () => {
    const { getByTestId } = renderWithTheme(
      <SidebarSection
        title={items.title}
        child={items.child}
        setNodes={jest.fn()}
      />
    );
    const button = getByTestId("sidebar-subsection-button");
    expect(button).toBeInTheDocument();
  });
  it("checks if the collapse tab is open when the button is clicked", () => {
    const { getByTestId } = renderWithTheme(
      <SidebarSection
        title={items.title}
        child={items.child}
        setNodes={jest.fn()}
      />
    );
    const sectionButton = getByTestId("sidebar-subsection-button");
    const collapseButton = getByTestId("sidebar-subsection-collapse");

    expect(collapseButton).toHaveAttribute("aria-hidden", "true");

    fireEvent.click(sectionButton);

    expect(collapseButton).toHaveAttribute("aria-hidden", "false");
  });
});
