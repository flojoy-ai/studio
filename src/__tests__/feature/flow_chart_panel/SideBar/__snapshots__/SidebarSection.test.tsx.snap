// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarSection matches snapshot 1`] = `
<div>
  <button
    class="mantine-UnstyledButton-root mantine-mp73ef"
    type="button"
  >
    <div
      class="mantine-Group-root mantine-18amkhr"
    >
      <div
        class="mantine-1a78l0v"
      >
        <div
          class="mantine-1l4yfk"
        >
          test
        </div>
      </div>
      <svg
        class="mantine-q42rgi"
        fill="none"
        height="14"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.5"
        style="transform: none;"
        viewBox="0 0 24 24"
        width="14"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
          stroke="none"
        />
        <polyline
          points="9 6 15 12 9 18"
        />
      </svg>
    </div>
  </button>
  <div
    aria-hidden="true"
    class="mantine-1avyp1d"
    style="box-sizing: border-box; display: none; height: 0px; overflow: hidden;"
  >
    <div
      style="opacity: 0; transition: opacity 200ms ease;"
    >
      <div
        class="mantine-Text-root mantine-1jjayxu"
      >
        <button
          class="mantine-UnstyledButton-root mantine-1pv9c1o"
          data-testid="sidebar-subsection-button"
          type="button"
        >
          <div
            class="mantine-Group-root mantine-18amkhr"
          >
            <div
              class="mantine-1a78l0v"
            >
              <div
                class="mantine-1l4yfk"
              >
                test
              </div>
            </div>
            <svg
              class="mantine-q42rgi"
              fill="none"
              height="14"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
              style="transform: none;"
              viewBox="0 0 24 24"
              width="14"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
                stroke="none"
              />
              <polyline
                points="9 6 15 12 9 18"
              />
            </svg>
          </div>
        </button>
        <div
          aria-hidden="true"
          class="mantine-1avyp1d"
          data-testid="sidebar-subsection-collapse"
          style="box-sizing: border-box; display: none; height: 0px; overflow: hidden;"
        >
          <div
            style="opacity: 0; transition: opacity 200ms ease;"
          >
            <div
              class="mantine-Text-root mantine-58236w"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
